Update local testnet doc and parameters

chong-he opened the pull request:
## Issue Addressed

1. #4536 
2. #4771 
3. el_bootnode not being killed when running `./stop_local_testnet.sh`

## Proposed Changes

An update to the doc: https://github.com/sigp/lighthouse/blob/stable/scripts/local_testnet/README.md on how to run local testnet

To fix 1, Revise `ETH1_FOLLOW_DISTANCE` and `SECONDS_PER_ETH1_BLOCK` to appropriate values.

To fix 2, updated the doc in [Manual creation of local testnet](https://github.com/sigp/lighthouse/tree/stable/scripts/local_testnet#manual-creation-of-local-testnet), and also fix the time using `setup_time.sh` (which is part of [start_local_testnet.sh](https://github.com/sigp/lighthouse/blob/stable/scripts/local_testnet/start_local_testnet.sh) so that the time in [genesis.json](https://github.com/sigp/lighthouse/blob/stable/scripts/local_testnet/genesis.json) is updated.

To fix 3, add `exec` to [el_bootnode.sh](https://github.com/sigp/lighthouse/blob/stable/scripts/local_testnet/el_bootnode.sh)

chong-he committed:
update

jimmygchen commented on scripts/local_testnet/README.md:
````diff
@@ -5,55 +5,99 @@ This setup can be useful for testing and development.
 make install-lcli
 ```
 
+This will build `lighthouse` and `lcli`. For `geth` and `bootnode`, go to [geth website](https://gethstore.blob.core.windows.net/builds/geth-alltools-linux-amd64-1.13.1-3f40e65c.tar.gz) and download the `Geth & Tools`. For example, to download and extract `Geth & Tools 1.13.1`:
````
Perhaps link this one instead (for "geth website") which has links to macos and windows downloads?
https://geth.ethereum.org/downloads

chong-he replied:
Fixed, thanks!

chong-he committed:
revise link

chong-he committed:
update parameters

chong-he commented:
After discussion with @michaelsproul , we revise `ETH1_FOLLOW_DISTANCE` to 128 and `SECONDS_PER_ETH1_BLOCK` to 3 and the local testnet is progressing without errors. Lower values of `ETH1_FOLLOW_DISTANCE` (e.g., 64 or lower) will cause a persistent WARN message:
```
WARN Syncing deposit contract block cache
```
So we revise `ETH1_FOLLOW_DISTANCE` to 128. 

For `SECONDS_PER_ETH1_BLOCK`, we set it equal to be the same as the consensus layer slot time

chong-he committed:
update doc

chong-he committed:
Update doc

jimmygchen commented on scripts/local_testnet/README.md:
````diff
@@ -5,55 +5,109 @@ This setup can be useful for testing and development.
 
+where `beacon_node_1` can be changed to `beacon_node_2`, `beacon_node_3` or `beacon_node_4` to view logs for different beacon nodes. The same applies to validator clients and geth nodes. 
+
+The view the debug logs for beacon nodes and validator clients:
+```
+tail -f ~/.lighthouse/local-testnet/node_1/beacon/logs/beacon.log
+tail -f ~/.lighthouse/local-testnet/node_1/validators/logs/validator.log
+```
+
+Once again, the `node_1` can be modified to view the debug logs for different nodes.
````
I think this is useful to know but may be confusing for new users trying to run testnet for the first time, having to look at logs at two different locations. There's also `-d DEBUG_LEVEL` with the `start_local_testnet.sh` -  which would make debug logs available if enabled, so maybe we can omit this?

chong-he replied:
> I think this is useful to know but may be confusing for new users trying to run testnet for the first time, having to look at logs at two different locations. There's also `-d DEBUG_LEVEL` with the `start_local_testnet.sh` - which would make debug logs available if enabled, so maybe we can omit this?

Good idea, I have removed them to avoid confusion

chong-he committed:
add quic port

chong-he committed:
remove debug log

chong-he committed:
Fix el_bootnode not being killed

jimmygchen commented on scripts/local_testnet/el_bootnode.sh:
```diff
@@ -1,3 +1,3 @@
 source ./vars.env
-$EL_BOOTNODE_BINARY --nodekeyhex $priv_key
\ No newline at end of file
+exec $EL_BOOTNODE_BINARY --nodekeyhex $priv_key
```
ahh nice find!

chong-he replied:
by @michaelsproul :)

chong-he committed:
Fix time

chong-he committed:
Fix doc in manual creation of testnet

chong-he committed:
Update file

chong-he commented:
For #4771, the fix is 2 parts:

- the doc is revised so that the beacon node is connecting to geth via the engine API port. The ports used in the example are revised so that they are consistent with the ports used by running `./start_local_testnet.sh genesis.json`

- the local testnet still couldn't progress via Merge/Capella. I believe this is due to the `shanghaiTime` in [genesis.json](https://github.com/sigp/lighthouse/blob/stable/scripts/local_testnet/genesis.json) not being updated:

https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/scripts/local_testnet/start_local_testnet.sh#L106-L111

So I created another file `setup_time.sh` (which is part of [start_local_testnet.sh](https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/scripts/local_testnet/start_local_testnet.sh)) so that the time in [genesis.json](https://github.com/sigp/lighthouse/blob/stable/scripts/local_testnet/genesis.json) is updated.  I don't know if there is a better / more elegant way to do this (i.e., adding another `.sh` file to update the [genesis.json](https://github.com/sigp/lighthouse/blob/stable/scripts/local_testnet/genesis.json)), feel free to comment 

After testing, manual creation of local testnet works and can progress well. However, initially, there are some errors in geth:

```
WARN Served engine_forkchoiceUpdatedV2   conn=127.0.0.1:41428 reqid=1 duration="434.925µs" err="Invalid parameters"   errdata="{Error:missing withdrawals list}"
```

and beacon node (believe to be related to the error above):

```
WARN Execution engine call failed error: ServerMessage { code: -32602, message: "Invalid parameters" }, service: exec
WARN Error whilst processing payload status error: Api { error: ServerMessage { code: -32602, message: "Invalid parameters" } }, service: exec
```

These error logs lasted for a minute or so, disappeared by itself and never returned. I tend to think this is ok, but if it's not, let me know.

chong-he committed:
update api doc

chong-he committed:
Revert "update api doc"

This reverts commit ed695743deca4818ad074e5285e04764965b1be5.

jimmygchen commented on scripts/local_testnet/setup_time.sh:
```diff
@@ -0,0 +1,14 @@
+
+source ./vars.env
+
+genesis_file=${@:$OPTIND+0:1}
```
`OPTIND` doesn't do anything here as we're not using the `getopts` command to parse command line arguments, so this can be simplified to:

```suggestion
genesis_file=$1
```

chong-he replied:
thank you!

jimmygchen commented on scripts/local_testnet/setup_time.sh:
```diff
@@ -0,0 +1,14 @@
+# Update future hardforks time in the EL genesis file based on the CL genesis time
+GENESIS_TIME=$(lcli pretty-ssz --spec $SPEC_PRESET --testnet-dir $TESTNET_DIR BeaconState $TESTNET_DIR/genesis.ssz | jq | grep -Po 'genesis_time": "\K.*\d')
+echo $GENESIS_TIME
+CAPELLA_TIME=$((GENESIS_TIME + (CAPELLA_FORK_EPOCH * 32 * SECONDS_PER_SLOT)))
```
I've just noticed that this won't work for `minimal` preset (`SLOTS_PER_EPOCH: 8`) and `gnosis` preset (`SLOTS_PER_EPOCH: 16`). I'm not sure if there's a nice way of getting the preset values (it's available via [beacon API](https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Config/getSpec) but requires beacon node running), perhaps we could just do a simple match the  `$SPEC_PRESET`?

```bash
get_spec_preset_value() {
  case "$SPEC_PRESET" in
    mainnet)   echo 32 ;;
    minimal)   echo 8  ;;
    gnosis)    echo 16 ;;
    *)         echo "Unsupported preset: $SPEC_PRESET" >&2; exit 1 ;;
  esac
}
```

chong-he replied:
Great catch, thank you so much!
Fix in [f629452](https://github.com/sigp/lighthouse/pull/4749/commits/f62945227a447bb3969f3323dcd88f179f624b56)

jimmygchen commented on scripts/local_testnet/setup_time.sh:
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
```
Could we replace the existing code in `start_local_testnet.sh` with a call to this script?
It may look like a bit of over-optimisation from first sight, but could avoid some maintenance issues later down the track, e.g. this would prevent us from forgetting to update `cancunTime` on the `Deneb` branch.

chong-he replied:
Do you mean adding the command in `setup_time.sh` to `start_local_testnet.sh`? Because `start_local_testnet.sh` is to automatically start the local testnet, while the `setup_time.sh` is to manually start it. 

I do agree that the `setup_time.sh` should be removed to keep it clean. How about merging `setup_time.sh` into `setup.sh`?

jimmygchen replied:
I meant to replace these lines in `start_local_testnet.sh`
https://github.com/sigp/lighthouse/blob/c0b6b92f2704198a2301fdbda0ca08c29ce7e99b/scripts/local_testnet/start_local_testnet.sh#L106-L110
with 
```
./setup_time.sh
```

So that we keep the "set up time" logic in one place rather than two. `unstable` now contains the Cancun changes, so you'll notice that the following would need to be updated in both places before the refactor:

```
CANCUN_TIME=$((GENESIS_TIME + (DENEB_FORK_EPOCH * 32 * SECONDS_PER_SLOT)))
echo $CANCUN_TIME
sed -i 's/"cancunTime".*$/"cancunTime": '"$CANCUN_TIME"',/g' $genesis_file
cat $genesis_file
```

chong-he replied:
Thank you for the comment
I have replaced the commands with  `./setup_time.sh genesis.json` in the file `start_local_testnet.sh` in [cbc3de1](https://github.com/sigp/lighthouse/pull/4749/commits/cbc3de1a3e71d8f3f9b677f15f6d8a8511ad5900)
This also means we have to move the `case` function in https://github.com/sigp/lighthouse/pull/4749#discussion_r1343981838 to `setup_time.sh`

For now:
-To start the local testnet automatically, we run `./start_local_testnet.sh genesis.json`
-To start the local testnet manually, we run a few commands, including `./setup_time.sh genesis.json`

Does it look strange that the `genesis.json` has now to be called by both `start_local_testnet.sh` and `setup_time.sh`?

chong-he replied:
Also added the CANCUN_TIME code in `setup_time.sh` in [cb74726](https://github.com/sigp/lighthouse/pull/4749/commits/cb74726bf739403b6ce883155a523207eb42c4cf)

jimmygchen commented on scripts/local_testnet/README.md:
```diff
@@ -5,58 +5,105 @@ This setup can be useful for testing and development.
 ## Requirements
 
-The scripts require `lcli`, `lighthouse`, `geth`, `bootnode` to be installed on `PATH`.
+The scripts require `lcli`, `lighthouse`, `geth`, `bootnode` to be installed on `PATH` (run `echo $PATH` to view all `PATH` directories).
```
FYI `geth` and `bootnode`  just need to be available at the paths defined in `vars.env`, but yeah by default that would need to be on the $PATH, so we can leave this as it is to keep things simple.

chong-he replied:
Thanks for the info, this is due to this?
https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/scripts/local_testnet/vars.env#L2-L3

jimmygchen replied:
Yep!

jimmygchen commented on scripts/local_testnet/README.md:
````diff
@@ -5,58 +5,105 @@ This setup can be useful for testing and development.
-root of this repository, run:
+The first step is to install Rust and dependencies. Refer to the [Lighthouse Book](https://lighthouse-book.sigmaprime.io/installation-source.html#dependencies) for installation. We will also need [jq](https://jqlang.github.io/jq/), which can be installed with `sudo apt install jq`.
+
+Then, we clone the Lighthouse repository:
+```bash
+cd ~
+https://github.com/sigp/lighthouse.git
+cd lighthouse
+```
````
`git clone` is missing here, but I think the above link to "installation from source" covers this step, but I'm not fuss - we could either fix command or just leave it as it was before.

chong-he replied:
thank you, fix in [b70d7c6](https://github.com/sigp/lighthouse/pull/4749/commits/b70d7c642685914d4cc8f7a574ffe375655f3a8c)

jimmygchen commented on scripts/local_testnet/README.md:
````diff
@@ -5,58 +5,105 @@ This setup can be useful for testing and development.
+To start a testnet using the predetermined settings:
+
+```bash
+cd ~
+cd /lighthouse/scripts/local_testnet
````
I think the path should be:

```suggestion
cd ~
cd lighthouse/scripts/local_testnet
```

Although I think it's probably fine to just document commands that runs  "From the root of the repository" to keep things simple.

chong-he replied:
I miss a `.`, fix in [00f68f2](https://github.com/sigp/lighthouse/pull/4749/commits/00f68f22b786e570c1e9b7ae8c92a8ace862ddec), hopefully that's good now

jimmygchen commented on scripts/local_testnet/README.md:
```diff
@@ -5,58 +5,105 @@ This setup can be useful for testing and development.
 ## Stopping the testnet
 
-This is not necessary before `start_local_testnet.sh` as it invokes `stop_local_testnet.sh` automatically.
+To stop the testnet, navigate to the directory `cd /lighthouse/scripts/local_testnet`, then run the command:
```
Same as above, I think it's usually fine to refer to files using relative path to the repo

```suggestion
To stop the testnet, navigate to the directory `scripts/local_testnet`, then run the command:
```

or full repo if you prefer (but without the `/` at the beginning)

```suggestion
To stop the testnet, navigate to the directory `cd lighthouse/scripts/local_testnet`, then run the command:
```

chong-he replied:
Fix in [9b4a969](https://github.com/sigp/lighthouse/pull/4749/commits/9b4a9699420ca28bf43d1ea588d448249f3128ae)  using `~` just in case the terminal is in a different working directory

jimmygchen commented on scripts/local_testnet/README.md:
```diff
@@ -5,58 +5,105 @@ This setup can be useful for testing and development.
 ## Manual creation of local testnet
 
-These scripts are used by ./start_local_testnet.sh and may be used to manually
+In [Starting the testnet](./README.md#starting-the-testnet), the testnet is started automatically with predetermined parameters (database directory, ports used etc).  This section describes some modifications of the local testnet settings, e.g., you would like to change the database directory, or changing the ports used. 
```
```suggestion
In [Starting the testnet](./README.md#starting-the-testnet), the testnet is started automatically with predetermined parameters (database directory, ports used etc).  This section describes some modifications of the local testnet settings, e.g., changing the database directory, or changing the ports used. 
```

jimmygchen commented on scripts/local_testnet/README.md:
````diff
@@ -94,10 +141,16 @@ beacon node:
+```bash
+./geth.sh $HOME/.lighthouse/local-testnet/geth_2 7002 6002 5002 genesis.json
+./beacon_node.sh $HOME/.lighthouse/local-testnet/node_2 9002 9102 8002 http://localhost:5002 ~/.lighthouse/local-testnet/geth_2/geth/jwtsecret
+./validator_client.sh $HOME/.lighthouse/local-testnet/node_2 http://localhost:8002
````
if you want to run validators on a new VC, it would require generating additional keys for that, i.e. updating `vars.env` and start over.

jimmygchen replied:
as discussed, by default we generate keys for 4 VCs, so this would be fine

jimmygchen reviewed (changes requested):
Nice fixes and updates! I've added a few comments.

chong-he committed:
add git clone

chong-he committed:
Fix path

chong-he committed:
Fix path

chong-he committed:
Merge branch 'sigp:stable' into local-doc

chong-he committed:
Update scripts/local_testnet/setup_time.sh

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

chong-he committed:
Update scripts/local_testnet/README.md

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

chong-he committed:
Fix SLOT_PER_EPOCH that changes with mainnet or minimal

chong-he committed:
Embedded setup_time.sh in start_local_testnet.sh

chong-he committed:
fix slot per epoch constant

chong-he committed:
Add comment

chong-he committed:
Add CANCUN_TIME

chong-he committed:
Fix CANCUN_TIME constant 32 slots

chong-he committed:
Merge branch 'unstable' into local-doc

chong-he committed:
Correct typo

chong-he commented:
> Nice fixes and updates! I've added a few comments.

Thank you so much for the review. I have addressed all comments, tested with [c3ae1f3](https://github.com/sigp/lighthouse/pull/4749/commits/c3ae1f3bad2fc4f03e7164bd51298ded11e29987) and the local testnet looks good on both `mainnet` and `minimal`, and also added Deneb/Cancun update to the scripts

jimmygchen committed:
chmod +x ./setup_time.sh

jimmygchen reviewed (approved):
Nice, looks good to me 👍 
I've noticed an issue on CI and fixed the `setup_time.sh` file permission, happy to merge once CI passes 🤞

paulhauner commented:
bors r+

paulhauner commented:
Hmm.. Bors doesn't seem to have detected that r+. I'll try again 🤞

bors r+

paulhauner commented:
It seems that bors is down. Trying again to see if it's back yet.

bors r+
